api/v1alpha6/conversion_test.go

```diff
@@ -19,29 +19,98 @@ package v1alpha6
 import (
 	"testing"
 
+	"github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
+	ctrlconversion "sigs.k8s.io/controller-runtime/pkg/conversion"
 
 	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha7"
 )
 
 func TestFuzzyConversion(t *testing.T) {
+	// The test already ignores the data annotation added on up-conversion.
+	// Also ignore the data annotation added on down-conversion.
+	ignoreDataAnnotation := func(hub ctrlconversion.Hub) {
+		obj := hub.(metav1.Object)
+		delete(obj.GetAnnotations(), utilconversion.DataAnnotation)
+	}
+
 	t.Run("for OpenStackCluster", utilconversion.FuzzTestFunc(utilconversion.FuzzTestFuncInput{
-		Hub:   &infrav1.OpenStackCluster{},
-		Spoke: &OpenStackCluster{},
+		Hub:              &infrav1.OpenStackCluster{},
+		Spoke:            &OpenStackCluster{},
+		HubAfterMutation: ignoreDataAnnotation,
 	}))
 
 	t.Run("for OpenStackClusterTemplate", utilconversion.FuzzTestFunc(utilconversion.FuzzTestFuncInput{
-		Hub:   &infrav1.OpenStackClusterTemplate{},
-		Spoke: &OpenStackClusterTemplate{},
+		Hub:              &infrav1.OpenStackClusterTemplate{},
+		Spoke:            &OpenStackClusterTemplate{},
+		HubAfterMutation: ignoreDataAnnotation,
 	}))
 
 	t.Run("for OpenStackMachine", utilconversion.FuzzTestFunc(utilconversion.FuzzTestFuncInput{
-		Hub:   &infrav1.OpenStackMachine{},
-		Spoke: &OpenStackMachine{},
+		Hub:              &infrav1.OpenStackMachine{},
+		Spoke:            &OpenStackMachine{},
+		HubAfterMutation: ignoreDataAnnotation,
 	}))
 
 	t.Run("for OpenStackMachineTemplate", utilconversion.FuzzTestFunc(utilconversion.FuzzTestFuncInput{
-		Hub:   &infrav1.OpenStackMachineTemplate{},
-		Spoke: &OpenStackMachineTemplate{},
+		Hub:              &infrav1.OpenStackMachineTemplate{},
+		Spoke:            &OpenStackMachineTemplate{},
+		HubAfterMutation: ignoreDataAnnotation,
 	}))
 }
+
+// Test that mutation of a converted object survives a subsequent conversion.
+func TestMutation(t *testing.T) {
```

Comment:
@lentzi90 This is the other test I was talking about which ensures that we don't simply overwrite an object with the stashed version on conversion.